src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,576 @@
+                            predicates.push(
+                                MirScalarExpr::column(column)
+                                    .call_unary(UnaryFunc::IsNull)
+                                    .call_unary(UnaryFunc::Not(expr::func::Not)),
+                            );
+                        }
+                    }
+                    predicates
+                } else {
+                    // Everything is true about an empty collection, but let's have other rules
+                    // optimize them away.
+
+                    Vec::new()
+                }
+            }
+            MirRelationExpr::Let { id, value, body } => {
+                // This deals with shadowed let bindings, but perhaps we should just complain instead?
+                let value_knowledge = PredicateKnowledge::action(value, let_knowledge)?;
+                let prior_knowledge =
+                    let_knowledge.insert(expr::Id::Local(id.clone()), value_knowledge);
+                let body_knowledge = PredicateKnowledge::action(body, let_knowledge)?;
+                let_knowledge.remove(&expr::Id::Local(id.clone()));
+                if let Some(prior_knowledge) = prior_knowledge {
+                    let_knowledge.insert(expr::Id::Local(id.clone()), prior_knowledge);
+                }
+                body_knowledge
+            }
+            MirRelationExpr::Project { input, outputs } => {
+                let mut input_knowledge = PredicateKnowledge::action(input, let_knowledge)?;
+                // Need to
+                // 1. restrict to supported columns,
+                // 2. rewrite constraints using new column identifiers,
+                // 3. add equivalence constraints for repeated columns.
+                let mut remap = HashMap::new();
+                for (index, column) in outputs.iter().enumerate() {
+                    if !remap.contains_key(column) {
+                        remap.insert(column, index);
+                    }
+                }
+
+                // 1. restrict to supported columns,
+                input_knowledge
+                    .retain(|expr| expr.support().into_iter().all(|c| remap.contains_key(&c)));
+
+                // 2. rewrite constraints using new column identifiers,
+                for predicate in input_knowledge.iter_mut() {
+                    predicate.visit_mut(&mut |e| {
+                        if let MirScalarExpr::Column(c) = e {
+                            *c = remap[c];
+                        }
+                    })
+                }
+
+                // 3. add equivalence constraints for repeated columns.
+                for (index, column) in outputs.iter().enumerate() {
+                    if remap[column] != index {
+                        input_knowledge.push(
+                            MirScalarExpr::Column(remap[column])
+                                .call_binary(MirScalarExpr::Column(*column), expr::BinaryFunc::Eq),
```

Comment:
This is not right IMO. If we had

```
Constant (null)
Project (#0, #0)
Filter (#0 = #1)
```

whose result should be:

```
Constant
```

we would be lifting that `#0 = #1` is known to be true and hence the Filter would be treated as redundant and removed, resulting in:

```
Constant (null, null)
```

instead.

Reply:
Same applies to join equivalences. In both cases, equivalences of this type should be lifted as `(e1 = e2) || (e1 is null and e2 is null)`.

Reply:
I''m not sure this is the case (the one about duplicated columns does seem right though). Join equivalences are "Rust Equality" rather than "SQL Equality". They do equate nulls.

Reply:
Oh, it seems `Eq` predicates are treated differently.

Reply:
Wait, actually. I might have it backwards. The predicates that are introduced are introduced with SQL equality, and you have the right predicate written up there. I got myself a bit confused because I thought a different problem was at work. I'm currently of the opinion that you are right in both cases (that the two "things", columns or join equiv arguments, are "Rust Equals", but the expression will be "SQL Equals" and so we need to have the `||` in there.

Sorry for the confusion!

Reply:
It turns out the wrong results issue was actually in this line, but it's not the one I was trying to describe above.

The wrong results issue spotted by https://buildkite.com/materialize/tests/builds/24125#9fecce25-f73b-436e-a19c-f036333ed790 was caused by this:

```
diff --git a/src/transform/src/predicate_propagation.rs b/src/transform/src/predicate_propagation.rs
index 8199308a8..74380e786 100644
--- a/src/transform/src/predicate_propagation.rs
+++ b/src/transform/src/predicate_propagation.rs
@@ -184,7 +184,7 @@ impl PredicateKnowledge {
                     if remap[column] != index {
                         input_knowledge.push(
                             MirScalarExpr::Column(remap[column])
-                                .call_binary(MirScalarExpr::Column(*column), expr::BinaryFunc::Eq),
+                                .call_binary(MirScalarExpr::Column(index), expr::BinaryFunc::Eq),
                         );
                     }
                 }
```

`column` is the offset before the projection, `index` is the new offset of the column.